### ISSUE TITLE
Severe performance regression of the Engine after Comparator API

### ISSUE THREAD

**Issue by Akirathan** [enso-org/enso#5636]
After #4067 was merged, there is a **severe performance regression** of the Engine. Looking at some benchmarks, they seem to regress by roughly 100x. The benchmarking jobs are even failing with `Error: The operation was cancelled`, maybe because the jobs takes too long now. An example of a recent successful benchmark job is https://github.com/enso-org/enso/actions/runs/4134490767/jobs/7145690994, and the failing job is https://github.com/enso-org/enso/actions/runs/4142523782/jobs/7163290560. Note that all the Engine-benchmark jobs since that time are failing.

We should investigate what is happening in these benchmarks and why the performance is so slow ASAP.

**Comment by hubertp** [comment]
See points 10 and 11 to quickly scan for deoptimizations https://github.com/enso-org/enso/blob/develop/docs/runtime-guide.md#tips-and-tricks


<blockquote><img src="https://opengraph.githubassets.com/7910deb6bbb59df25e2b0d6557a524c60156c4698709fb344ee8f153fae7f663/enso-org/enso" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">enso/runtime-guide.md at develop · enso-org/enso</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

**Comment by Akirathan** [comment]
Analyzing performance of `Any.==`. Comparing a revision before Comparator API (https://github.com/enso-org/enso/commit/4f70bcc5ff42a898cc61a83e8daae988f11d6f57) [_old_] with most recent revision (6971dd08506adf0cb07c0f7da1cc859b2137e7bc) [_new_]. The source is:
```
from Standard.Base import all

my_func x =
    x == True

main =
    n = 100000
    iter_size = 100

    integer_vec = Vector.new n (ix -> ix)
    boolean_vec = integer_vec.map (x-> x % 5 == 0)

    res = 0.up_to iter_size . map _->
        boolean_vec.map bool->
            my_func bool

    res.first.first
```

Run with:
```
env JAVA_OPTS='-Dpolyglot.engine.CompilationStatistics=true -Dpolyglot.engine.TracePerformanceWarnings=all -Dpolyglot.engine.TraceCompilationDetails=true -Dpolyglot.engine.TraceInliningDetails=true' enso --no-ir-caches --log-level INFO --run tmp.enso
```

These are the outputs:
- [new.txt](https://github.com/enso-org/enso/files/10725314/new.txt)
- [old.txt](https://github.com/enso-org/enso/files/10725315/old.txt)

At first sight, there are no performance warnings neither in old, neither in new. In new, everything in `Any.==` is inlined, as expected. Looking at the compilation statistics, _new_ has 3x more compilations (151) than _old_ (46). So far, I have not found a simple bottleneck, it just seems that there is just more Enso code to compile.

**Comment by wdanilo** [comment]
@Akirathan why the regressions were not discovered before merging PR? We should always run benchmarks before PR is merged - what happened here?

**Comment by Akirathan** [comment]
@wdanilo There was a lot of pressure to finalize #4067, as it should be the last API-breaking change from the engine side. With @JaroslavTulach, we decided that we want to merge it as soon as possible, and deal with all the performance regressions as follow-up.

**Comment by Akirathan** [comment]
@wdanilo Moreover, benchmarks are run as post-merge jobs. Usually, this job takes several hours, so running it as a pre-merge check would not be a good idea. In the engine team, we run benchmarks locally, if we know that there might be some regressions. I just have not done that as part of #4067, because of the time pressure, and postponed it as a follow-up.

**Comment by wdanilo** [comment]
@Akirathan gotcha. How much time does it take to run benchmarks locally? I still think we should run them for all the PRs. Merging a PR even couple hours later is probably better than not discovering perf regressions. What do you think about it?

**Comment by JaroslavTulach** [comment]
> ... a lot of pressure to finalize #4067, as... @JaroslavTulach, ... decided that we want to merge it as soon as possible, 
> and deal with all the performance regressions as follow-up.

It is my decision and my fault. Pavel did a magnificent work of providing _hashcode_ support for all Enso objects. It took more than two months and (even when split to multiple sprints and pivotals) it turned out to be too huge. At the end I decided to not wait anymore, integrate and make sure all the _API changes_ are in.

The regression is unfortunate, but it shall not affect ability of other teams to more forward - the API is there. We just need to find out what's wrong in the engine and speed it up - it is not going to be a non-solvable issue - we shall be able to get on par with speed of the previous behavior in a week or two. We'll have a pair programming with Pavel on Thursday - we'll focus on finding the cause.

**Comment by wdanilo** [comment]
@JaroslavTulach oh, no worries. Look, I was only worried that we can in the future overlook some regressions if we are not running benchmarks with every PR. That's my motivation behind asking these questions. I'm happy that we discovered the regression here, but if we are not running benchmarks before merging the a new code, we might (and probably will) not discover similar cases in the future. Thanks @Akirathan and @JaroslavTulach for taking care of it! ❤️

**Comment by Akirathan** [comment]
> @JaroslavTulach oh, no worries. Look, I was only worried that we can in the future overlook some regressions if we are not running benchmarks with every PR. That's my motivation behind asking these questions. I'm happy that we discovered the regression here, but if we are not running benchmarks before merging the a new code, we might (and probably will) not discover similar cases in the future. Thanks @Akirathan and @JaroslavTulach for taking care of it! heart

I am afraid running benchmarks as part of pre-merge checks is simply not possible. Currently, engine benchmarking job takes roughly 3 hours. But not only that, I suppose that the job has some special scheduling policies (at least they should) - for example, you may want to schedule a benchmarking job on a node without any other jobs, so that stuff like HW perf counters work as expected, and are not interfered with. Moreover, we currently have just _one_ benchmarking job taking 3 hours, these are just _micro_ benchmarks, I expect this number to grow significantly in the future.

**Comment by JaroslavTulach** [comment]
I tried to verify and while an visible improvement is there, I had to report: 
- #5712

as a needed continuation of the effort.
